src/vs/workbench/contrib/notebook/browser/viewParts/notebookEditorToolbar.ts

```diff
@@ -1174,3 +617,105 @@ export function workbenchCalculateActions(initialPrimaryActions: IActionModel[],
+		return { primaryActions: [], secondaryActions: initialSecondaryActions };
+	}
+
+	// find true length of array, add 1 for each primary actions, ignoring an item when size = 0
+	const visibleActionLength = initialPrimaryActions.filter(action => action.size !== 0).length;
+
+	// step 1: try to fit all primary actions
+	const totalWidthWithLabels = initialPrimaryActions.map(action => action.size).reduce((a, b) => a + b, 0) + (visibleActionLength - 1) * ACTION_PADDING;
+	if (totalWidthWithLabels <= leftToolbarContainerMaxWidth) {
+		initialPrimaryActions.forEach(action => {
+			action.renderLabel = true;
+		});
+		return {
+			primaryActions: initialPrimaryActions.map(action => action.action),
+			secondaryActions: initialSecondaryActions
+		};
+	}
+
+	// step 2: check if they fit without labels
+	if ((visibleActionLength * ICON_ONLY_ACTION_WIDTH + (visibleActionLength - 1) * ACTION_PADDING) > leftToolbarContainerMaxWidth) {
+		return calculateIconOverflow(initialPrimaryActions, initialSecondaryActions, leftToolbarContainerMaxWidth);
+	}
+
+	// step 3: render as many actions as possible with labels, rest without.
+	let sum = 0;
+	let lastActionWithLabel = -1;
+	for (let i = 0; i < initialPrimaryActions.length; i++) {
+		sum += initialPrimaryActions[i].size + ACTION_PADDING;
+
+		if (initialPrimaryActions[i].action instanceof Separator) {
+			// find group separator
+			const remainingItems = initialPrimaryActions.slice(i + 1);
+			const newTotalSum = sum + (remainingItems.length === 0 ? 0 : (remainingItems.length * ICON_ONLY_ACTION_WIDTH + (remainingItems.length - 1) * ACTION_PADDING));
+			if (newTotalSum <= leftToolbarContainerMaxWidth) {
+				lastActionWithLabel = i;
+			}
+		} else {
+			continue;
+		}
+	}
+
+	// icons only don't fit either
+	// todo: don't quite understand this bit
+	if (lastActionWithLabel < 0) {
+		return calculateIconOverflow(initialPrimaryActions, initialSecondaryActions, leftToolbarContainerMaxWidth);
+	}
+
+
+	initialPrimaryActions.slice(0, lastActionWithLabel + 1).forEach(action => { action.renderLabel = true; });
+	initialPrimaryActions.slice(lastActionWithLabel + 1).forEach(action => { action.renderLabel = false; });
+	return {
+		primaryActions: initialPrimaryActions.map(action => action.action),
+		secondaryActions: initialSecondaryActions
+	};
+}
+
+function calculateIconOverflow(initialPrimaryActions: any[], initialSecondaryActions: any[], leftToolbarContainerMaxWidth: number) {
+	initialPrimaryActions.forEach(action => { action.renderLabel = false; });
+	const renderActions: IActionModel[] = [];
+	const overflow: IAction[] = [];
+
+	let currentSize = 0;
+	for (let i = 0; i < initialPrimaryActions.length; i++) {
+		const actionModel = initialPrimaryActions[i];
+		const itemSize = ICON_ONLY_ACTION_WIDTH;
+
+		if (currentSize + itemSize <= leftToolbarContainerMaxWidth) {
+			currentSize += ACTION_PADDING + itemSize;
+			renderActions.push(actionModel);
+		} else {
+			overflow.push(actionModel.action);
```

Comment:
If actionModel.size === 0, then we should append it to `renderActions` other than `overflow`, like what you did in https://github.com/microsoft/vscode/blob/80ffe8ebdcb9828e59bea511ed39164026bab0f4/src/vs/workbench/contrib/notebook/browser/viewParts/notebookEditorToolbar.ts#L583-L585